README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggpmisc <img src="man/figures/logo-ggpmisc.png" align="right" width="150" />

[![cran
version](https://www.r-pkg.org/badges/version/ggpmisc)](https://cran.r-project.org/package=ggpmisc)
[![R build
status](https://github.com/aphalo/ggpmisc/workflows/R-CMD-check/badge.svg)](https://github.com/aphalo/ggpmisc/actions)

## Purpose

Package ‘**ggpmisc**’ (Miscellaneous Extensions to ‘ggplot2’) is a set
of extensions to R package ‘ggplot2’ (&gt;= 3.0.0) with emphasis on
annotations and highlighting related to fitted models and data
summaries. Data summaries shown as text, tables or equations are
implemented. New geoms support insets in ggplots.

## Extended Grammar of graphics

The position of annotations within the plotting area depends in most
cases on graphic design considerations rather than on properties such as
the range of values in the data being plotted. In particular, the
location within the plotting area of large annotation objects like
model-fit summaries, location maps, plots, and tables needs usually to
be set independently of the `x` and `y` scales, re-scaling or any
transformations. To acknowledge this, the Grammar of Graphics is here
expanded by supporting *x* and *y* positions expressed in ‘grid’ “npc”
units in the range 0..1. This is implemented with new
(pseudo-)aesthetics *npcx* and *npcy* and their corresponding scales,
plus geometries and a revised `annotate()` function. The new aesthetics
function in “parallel” with the *x* and *y* aesthetics used for plotting
data. The advantage of this approach is that the syntax used for
annotations becomes identical to that used for plotting data and that
annotations with approach *cleanly* support facets in a way consistent
with the rest of the grammar.

## Aesthetics and scales

Scales `scale_npcx_continuous()` and `scale_npcy_continuous()` and the
corresponding new aesthetics `npcx` and `npcy` make it possible to add
graphic elements and text to plots using coordinates expressed in `npc`
units for the location within the plotting area.

Scales `scale_x_logFC()` and `scale_y_logFC()` are suitable for plotting
of log fold change data. Scales `scale_x_Pvalue()`, `scale_y_Pvalue()`,
`scale_x_FDR()` and `scale_y_FDR()` are suitable for plotting *p*-values
and adjusted *p*-values or false discovery rate (FDR). Default arguments
are suitable for volcano and quadrant plots as used for transcriptomics,
metabolomics and similar data.

Scales `scale_colour_outcome()`, `scale_fill_outcome()` and
`scale_shape_outcome()` and functions `outome2factor()`,
`threshold2factor()`, `xy_outcomes2factor()` and
`xy_thresholds2factor()` used together make it easy to map ternary
numeric outputs and logical binary outcomes to color, fill and shape
aesthetics. Default arguments are suitable for volcano, quadrant and
other plots as used for genomics, metabolomics and similar data.

## Geometries

Geometries `geom_table()`, `geom_plot()` and `geom_grob()` make it
possible to add inset tables, inset plots, and arbitrary ‘grid’
graphical objects including bitmaps and vector graphics as layers to a
ggplot using native coordinates for `x` and `y`.

Geometries `geom_text_npc()`, `geom_label_npc()`, `geom_table_npc()`,
`geom_plot_npc()` and `geom_grob_npc()`, `geom_text_npc()` and
`geom_label_npc()` are versions of geometries that accept positions on
*x* and *y* axes using aesthetics `npcx` and `npcy` values expressed in
“npc” units.

Geometries `geom_x_margin_arrow()`, `geom_y_margin_arrow()`,
`geom_x_margin_grob()`, `geom_y_margin_grob()`, `geom_x_margin_point()`
and `geom_y_margin_point()` make it possible to add marks along the *x*
and *y* axes. `geom_vhlines()` and `geom_quadrant_lines()` draw vertical
and horizontal reference lines within a single layer.

## Statistics

Statistic `stat_fmt_tb()` helps with the formatting of tables to be
plotted with `geom_table()`.

Statistics `stat_peaks()` and `stat_valleys()` can be used to highlight
and/or label maxima and minima in a plot.

Statistics that help with reporting the results of model fits are
`stat_poly_eq()`, `stat_fit_residuals()`, `stat_fit_deviations()`,
`stat_fit_glance()`, `stat_fit_augment()`, `stat_fit_tidy()` and
`stat_fit_tb()`.

Four statistics, `stat_dens2d_filter()`, `stat_dens2d_label()`,
`stat_dens1d_filter()` and `stat_dens1d_label()`, implement tagging or
selective labeling of observations based on the local 2D density of
observations in a panel. Another two statistics,
`stat_dens1d_filter_g()` and `stat_dens1d_filter_g()` compute the
density by group instead of by plot panel. These six stats are designed
to work well together with `geom_text_repel()` and `geom_label_repel()`
from package ‘ggrepel’.

A summary statistic using special grouping for quadrants
`stat_quadrant_counts()` can be used to automate labeling with the
number of observations.

The statistics `stat_apply_panel()` and `stat_apply_group()` can be
useful for applying arbitrary functions returning numeric vectors. They
are specially useful with functions lime `cumsum()`, `cummax()` and
`diff()`.

## ggplot methods

Being `ggplot()` defined as a generic method in ‘ggplot2’ makes it
possible to define specializations, and we provide two for time series
stored in objects of classes `ts` and `xts` which automatically convert
these objects into tibbles and set by default the aesthetic mappings for
`x` and `y` automatically. A companion function `try_tibble()` is also
exported.

## MIGRATED

Functions for the manipulation of layers in ggplot objects, together
with statistics and geometries useful for debugging extensions to
package ‘ggplot2’, earlier included in this package are now in package
‘gginnards’. [![cran
version](https://www.r-pkg.org/badges/version/gginnards)](https://cran.r-project.org/package=gginnards)

## Examples

``` r
library(ggpmisc)
library(ggrepel)
```

In the first example we plot a time series using the specialized version
of `ggplot()` that converts the time series into a tibble and maps the
`x` and `y` aesthetics automatically. We also highlight and label the
peaks using `stat_peaks`.

``` r
ggplot(lynx, as.numeric = FALSE) + geom_line() + 
  stat_peaks(colour = "red") +
  stat_peaks(geom = "text", colour = "red", angle = 66,
             hjust = -0.1, x.label.fmt = "%Y") +
  stat_peaks(geom = "rug", colour = "red", sides = "b") +
  expand_limits(y = 8000)
```

![](man/figures/README-readme-03-1.png)<!-- -->

In the second example we add the equation for a fitted polynomial plus
the adjusted coefficient of determination to a plot showing the
observations plus the fitted curve, deviations and confidence band. We
use `stat_poly_eq()`.

``` r
formula <- y ~ x + I(x^2)
ggplot(cars, aes(speed, dist)) +
  geom_point() +
  stat_fit_deviations(method = "lm", formula = formula, colour = "red") +
  geom_smooth(method = "lm", formula = formula) +
  stat_poly_eq(aes(label =  paste(stat(eq.label), stat(adj.rr.label), sep = "*\", \"*")),
               formula = formula, parse = TRUE)
```

![](man/figures/README-readme-04-1.png)<!-- -->

The same figure as in the second example but this time annotated with
the ANOVA table for the model fit. We use `stat_fit_tb()` which can be
used to add ANOVA or summary tables.

``` r
formula <- y ~ x + I(x^2)
ggplot(cars, aes(speed, dist)) +
  geom_point() +
  geom_smooth(method = "lm", formula = formula) +
  stat_fit_tb(method = "lm",
              method.args = list(formula = formula),
              tb.type = "fit.anova",
              tb.vars = c(Effect = "term", 
                          "df",
                          "M.S." = "meansq", 
                          "italic(F)" = "statistic", 
                          "italic(P)" = "p.value"),
              tb.params = c(x = 1, "x^2" = 2),
              label.y.npc = "top", label.x.npc = "left",
              size = 2.5,
              parse = TRUE)
#> Dropping params/terms (rows) from table!
```

![](man/figures/README-readme-05-1.png)<!-- -->

A plot with an inset plot.

``` r
p <- ggplot(mtcars, aes(factor(cyl), mpg, colour = factor(cyl))) +
  stat_boxplot() +
  labs(y = NULL) +
  theme_bw(9) + theme(legend.position = "none")

ggplot(mtcars, aes(wt, mpg, colour = factor(cyl))) +
  geom_point() +
  annotate("plot_npc", npcx = "left", npcy = "bottom", label = p) +
  expand_limits(y = 0, x = 0)
```

![](man/figures/README-readme-06-1.png)<!-- -->

A quadrant plot with counts and labels, using `geom_text_repel()` from
package ‘ggrepel’.

``` r
ggplot(quadrant_example.df, aes(logFC.x, logFC.y)) +
  geom_point(alpha = 0.3) +
  geom_quadrant_lines() +
  stat_quadrant_counts() +
  stat_dens2d_filter(color = "red", keep.fraction = 0.02) +
  stat_dens2d_labels(aes(label = gene), keep.fraction = 0.02, 
                     geom = "text_repel", size = 2, colour = "red") +
  scale_x_logFC(name = "Transcript abundance after A%unit") +
  scale_y_logFC(name = "Transcript abundance after B%unit")
```

![](man/figures/README-unnamed-chunk-1-1.png)<!-- -->

## Installation

Installation of the most recent stable version from CRAN:

``` r
install.packages("ggpmisc")
```

Installation of the current unstable version from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("aphalo/ggpmisc")
```

## Documentation

HTML documentation is available at
(<https://docs.r4photobiology.info/ggpmisc/>), including a *User Guide*.

News about updates are regularly posted at
(<https://www.r4photobiology.info/>).

## Contributing

Please report bugs and request new features at
(<https://github.com/aphalo/ggpmisc/issues>). Pull requests are welcome
at (<https://github.com/aphalo/ggpmisc>).

## Citation

If you use this package to produce scientific or commercial
publications, please cite according to:

``` r
citation("ggpmisc")
#> 
#> To cite package 'ggpmisc' in publications use:
#> 
#>   Pedro J. Aphalo (2020). ggpmisc: Miscellaneous Extensions to
#>   'ggplot2'. https://docs.r4photobiology.info/ggpmisc/,
#>   https://github.com/aphalo/ggpmisc.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Manual{,
#>     title = {ggpmisc: Miscellaneous Extensions to 'ggplot2'},
#>     author = {Pedro J. Aphalo},
#>     year = {2020},
#>     note = {https://docs.r4photobiology.info/ggpmisc/,
#> https://github.com/aphalo/ggpmisc},
#>   }
```

## License

© 2016-2020 Pedro J. Aphalo (<pedro.aphalo@helsinki.fi>). Released under
the GPL, version 2 or greater. This software carries no warranty of any
kind.


NEWS
ggpmisc 0.3.8 (2020-12-21)
==========================

  * Add param p.digits to stat_fit_tb().
  * Vignette explaing how the grammar of graphics is expanded to support annotations.
  * Fix bug: try_tibble.ts() and try_data_frame() did not handle correctly the
  conversion of dates for some time series, which also could affect ggplot.ts().
  * Fix bug: stat_peaks() and stat_valleys() generated wrong labels if a Date object was mapped to x.
  * Move git repository from Bitbucket to Github.

ggpmisc 0.3.7 (2020-11-09)
==========================

  * Update stat_fit_tb() to support renaming of terms/parameter names in the
  table (Suggested by Big Old Dave and Z. Lin). In addition implement selection,
  reordering and renaming of columns and terms/parameters using positional indexes
  and pattern matching of truncated names in addition to whole names. Improve
  formatting of small P-values.
  * Update stat_fmt_tb() to support the same expanded syntax as stat_fit_tb().
  * Add stat_dens1d_filter(), stat_dens1d_filter_g() and stat_dens1d_labels(),
  to complement existing stat_dens2d_filter(), stat_dens2d_filter_g() and
  stat_dens2d_labels().
  * Update stat_dens2d_filter(), stat_dens2d_filter_g() and stat_dens2d_labels()
  adding formal parameters keep.sparse and invert.selection, as available in the
  new 1D versions.
  * Update stat_dens2d_labels() to accept not only character strings but also
  functions as argument to label.fill as the new stat_dens1d_labels() does.
  * Revise documentation including the User Guide.

ggpmisc 0.3.6 (2020-09-06)
==========================

  * Override ggplot2::annotate() adding support for aesthetics npcx and npcy.
  * Add stat_summary_xy() and stat_centroid().
  * Revise stat_poly_eq() to support labeling of equations according to group.
  * Implement output.type "markdown" in stat_poly_eq() usable with geom_richtext()
  from package 'ggtext'.

ggpmisc 0.3.5 (2020-06-01)
==========================

  * Add support for "table themes" to geom_table() and geom_table_npc().

ggpmisc 0.3.4 (2020-04-22)
==========================

  * Add support for p.value.label and f.value.label to stat_poly_eq().
  * Update to track deprecations in 'ggplot2' (>= 3.3.0).

ggpmisc 0.3.3 (2019-11-23)
==========================

  * Fix bug in stat_poly_eq().
  * Minor revision of the User Guide and documentation.

ggpmisc 0.3.2 (2019-11-12)
==========================

This version implements some new features and fixes bugs in the features
introduced in version 0.3.1, please do rise an issue if you notice any
remaining bugs! Some reported weaknesses in the documentation have been
addressed. This updated version depends on 'ggplot2' (>= 3.2.1).

  * Add support for volcano and quadrant plots of outcomes.
  * Add geometries geom_vhlines() and geom_quadrant_lines().
  * Add convenience scales scale_x_logFC() and scale_y_logFC() for data expressed
  as fold change.
  * Add convenience scales scale_x_Pvalue(), scale_y_Pvalue(), scale_x_FDR(),
  scale_y_FDR().
  * Add convenience scales scale_colour_outcome(), scale_fill_outcome() and
  scale_shape_outcome() for data expressed as ternary or binary outcomes.
  * Add conversion functions outcome2factor() and threshold2factor() to convert
  vectors of numeric outcomes into factors with 2 or 3 levels.
  * Add conversion function xy_outcomes2factor() and xy_thresholds2factor() to
  combine two vectors of numeric outcomes into a 4-level factor.

  * Improve support for model-fit annotations.
  * Update stat_poly_eq() so that optionally instead of text labels it can return
  numeric values extracted from the fit object.
  * Document with examples how to pass weights and covariates to statistics based
  on methods from package 'broom'. Highlight the differences among stat_poly_eq()
  and the stat_fit_xxx() statistics implemented using package 'broom'.

  * Revise stat_apply_fun() to allow simultaneous application of functions to
  x and y aesthetics, and handling of diff() and other functions returning
  slightly shorter vectors than their input.

  * Support in stat_fit_tb(), stat_fit_augment(), stat_fit_tidy() and
  stat_fit_glance() the use of character strings as position arguments for
  parameters label.x and label.y when using geoms based on x and y aesthetics in
  addition to when using those taking the npcx and npcy aesthetics.

ggpmisc 0.3.1 (2019-04-01)
==========================

This is a major update, with a few cases in which old code may need to be
revised to work, and many cases in which there will be subtle differences in
the positions of labels used as annotations. The many new features may still
have some bugs, please do rise an issue if you notice one!

Version requiring 'ggplot2' (>= 3.1.0).

Add new geometries, several of them accepting x and y in npc units through
the new aesthetics npcx and npcy, allowing positioning relative to plotting
area irrespective of native data units and scale limits. These geometries are
useful on their own for annotations in particular they allow consistent
positioning of textual summaries. By default they do not inherit the plot's
aesthetic mappings making their behaviour remain by default in-between that
of true geometries and that of annotate().

  * Add geom_text_npc() and geom_label_npc() using aesthetics npcx and npcy.
  * Add geom_table_npc() using aesthetics npcx and npcy.
  * Add geom_plot() and geom_plot_npc() which can be used to add inset plots
  to a ggplot.
  * Add geom_grob() and geom_grob_npc() which can be used to add inset grobs
  to a ggplot.
  * Add geom_x_margin_point(), geom_y_margin_point(), geom_x_margin_arrow() and
  geom_y_margin_arrow() which behave similarly to geom_hline() and geom_vline()
  but plot points or arrows instead of lines. Add geom_x_margin_grob() and
  geom_y_margin_grob() with similar behaviour but for adding grobs.
  - Revise textual-summary statistics to use the new npc version of geometries.

R/ggpmisc.R
#' @details The new facilities for cleanly defining new stats and geoms added to
#' 'ggplot2' in version 2.0.0 and the support for nested tibbles and new syntax
#' for mapping computed values to aesthetics added to
#' 'ggplot2' in version 3.0.0 are used in this package's code. This
#' means that 'ggpmisc' (>= 0.3.0) requires version 3.0.0 or later of ggplot2
#' while 'ggpmisc' (< 0.3.0) requires version 2.0.0 or later of ggplot2.
#'
#' Extensions provided:
#' \itemize{
#' \item Function for conversion of time series data into tibbles that can be
#' plotted with ggplot.
#' \item \code{ggplot()} method for time series data.
#' \item Stats for locating and tagging "peaks" and "valleys" (local or global
#'  maxima and minima).
#' \item Stat for generating labels from a \code{lm()} model fit, including
#' formatted equation. By default labels are expressions but tikz device is
#' supported optionally with LaTeX formatted labels.
#' \item Stats for extracting information from a any model fit supported by
#' package 'broom'.
#' \item Stats for filtering-out/filtering-in observations in regions of a
#' panel or group where the density of observations is high.
#' \item Geom for annotating plots with tables.
#' }
#'
#' The stats for peaks and valleys are coded so as to work correctly both with
#' numeric and POSIXct variables mapped to the x aesthetic. Special handling was
#' needed as text labels are generated from the data.
#'
#' @section Warning!: \code{geom_null()}, \code{stat_debug_group()},
#'   \code{stat_debug_panel()}, \code{geom_debug()}, \code{append_layers()},
#'   \code{bottom_layer()}, \code{delete_layers()}, \code{extract_layers()},
#'   \code{move_layers()}, \code{num_layesr()}, \code{shift_layers()},
#'   \code{top_layer()} and \code{which_layers()} have been moved from package
#'   'ggpmisc' into their own seperate package
#'   '\code{\link[gginnards]{gginnards-package}}.
#'
#' @section Acknowledgements: We thank Kamil Slowikowski not only for
#'   contributing ideas and code examples to this package but also for adding
#'   new features to his package 'ggrepel' that allow new use cases for
#'   \code{stat_dens2d_labels} from this package.
#'
#' @references Package suite 'r4photobiology' web site at
#' \url{https://www.r4photobiology.info/}\cr Package 'ggplot2' documentation at
#' \url{https://ggplot2.tidyverse.org/}\cr Package 'ggplot2' source code at
#' \url{https://github.com/tidyverse/ggplot2}
#'
#' @import scales grid ggplot2
#' @importFrom ggplot2 ggplot
#' @importFrom magrittr %>%
#' @importFrom rlang .data
#'
#' @note The signatures of \code{stat_peaks()} and \code{stat_valleys()} are
#'   identical to those of \code{stat_peaks} and \code{stat_valleys} from
#'   package \code{photobiology} but the variables returned are a subset as
#'   values related to light spectra are missing. Furthermore the stats from
#'   package \code{ggpmisc} work correctly when the x aesthetic uses a date or
#'   datetime scale, while those from package \code{photobiology} do not
#'   generate correct labels in this case.
#'
#' @examples
#' library(tibble)
#'
#' ggplot(lynx, as.numeric = FALSE) + geom_line() +
#' stat_peaks(colour = "red") +
#'   stat_peaks(geom = "text", colour = "red", angle = 66,
#'              hjust = -0.1, x.label.fmt = "%Y") +
#'   ylim(NA, 8000)
#'
#' formula <- y ~ poly(x, 2, raw = TRUE)
#' ggplot(cars, aes(speed, dist)) +
#'   geom_point() +
#'   geom_smooth(method = "lm", formula = formula) +
#'   stat_poly_eq(aes(label = stat(eq.label)),
#'                formula = formula,
#'                parse = TRUE) +
#'   labs(x = expression("Speed, "*x~("mph")),
#'        y = expression("Stopping distance, "*y~("ft")))
#'
#' formula <- y ~ x
#' ggplot(PlantGrowth, aes(group, weight)) +
#'   stat_summary(fun.data = "mean_se") +
#'   stat_fit_tb(method = "lm",
#'               method.args = list(formula = formula),
#'               tb.type = "fit.anova",
#'               tb.vars = c(Term = "term", "df", "M.S." = "meansq",
#'                           "italic(F)" = "statistic",
#'                           "italic(p)" = "p.value"),
#'               tb.params = c("Group" = 1, "Error" = 2),
#'               table.theme = ttheme_gtbw(parse = TRUE)) +
#'   labs(x = "Group", y = "Dry weight of plants") +
#'   theme_classic()
#'
"_PACKAGE"
